packages/dev/inspector-v2/src/components/properties/particles/particleSystemProperties.tsx

```diff
@@ -66,44 +1066,216 @@ export const ParticleSystemColorProperties: FunctionComponent<{ particleSystem:
         </>
     );
 };
 
-// TODO-iv2: This can be more generic to work for not just particleSystems
-const useParticleSystemProperty = (
-    system: ParticleSystem,
-    propertyKey: keyof ParticleSystem,
-    observableType: "function" | "property",
-    addFn: keyof ParticleSystem,
-    removeFn: keyof ParticleSystem,
-    changeFn?: keyof ParticleSystem
-) => {
+/**
+ * Display spritesheet-related properties for a particle system.
+ * @param props Component props.
+ * @returns Render property lines.
+ */
+export const ParticleSystemSpritesheetProperties: FunctionComponent<{ particleSystem: ParticleSystem }> = (props) => {
+    const { particleSystem: system } = props;
+
+    return (
+        <>
+            <BoundProperty component={SwitchPropertyLine} label="Animation sheet enabled" target={system} propertyKey="isAnimationSheetEnabled" />
+            <BoundProperty component={NumberInputPropertyLine} label="First sprite index" target={system} propertyKey="startSpriteCellID" min={0} step={1} />
+            <BoundProperty component={NumberInputPropertyLine} label="Last sprite index" target={system} propertyKey="endSpriteCellID" min={0} step={1} />
+            <BoundProperty component={SwitchPropertyLine} label="Animation loop" target={system} propertyKey="spriteCellLoop" />
+            <BoundProperty component={SwitchPropertyLine} label="Random cell start index" target={system} propertyKey="spriteRandomStartCell" />
+            <BoundProperty component={NumberInputPropertyLine} label="Cell width" target={system} propertyKey="spriteCellWidth" min={0} step={1} />
+            <BoundProperty component={NumberInputPropertyLine} label="Cell height" target={system} propertyKey="spriteCellHeight" min={0} step={1} />
+            <BoundProperty component={NumberInputPropertyLine} label="Cell change speed" target={system} propertyKey="spriteCellChangeSpeed" min={0} step={0.01} />
+        </>
+    );
+};
+
+// Return a copied array and re-render when array mutators run.
+// Intercept add/remove/change functions because the underlying APIs update internal arrays in-place.
+const useObservableArray = <TargetT extends object, ItemT>(
+    target: TargetT,
+    getItems: () => ReadonlyArray<ItemT> | null | undefined,
+    addFn: keyof TargetT,
+    removeFn: keyof TargetT,
+    changeFn?: keyof TargetT
+): ItemT[] => {
     return useObservableState(
         useCallback(() => {
-            const value = observableType === "function" ? system[propertyKey]() : system[propertyKey];
-            return [...(value ?? [])];
-        }, [system, propertyKey]),
-        useInterceptObservable("function", system, addFn),
-        useInterceptObservable("function", system, removeFn),
-        changeFn ? useInterceptObservable("function", system, changeFn) : undefined
+            const value = getItems();
+            return [...(value ?? [])] as ItemT[];
+        }, [getItems]),
+        useInterceptObservable("function", target as any, addFn as any),

```

Comment:
Why are the `any` casts needed?

Reply:
That code was like that before. I'll see if I can remove them.